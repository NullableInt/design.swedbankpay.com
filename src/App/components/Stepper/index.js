import React from "react";
import PropTypes from "prop-types";
import classnames from "classnames";

const DefaultStepper = [
    {
        title: "Step one",
        subtitle: "24.12.17 12:10",
        completed: true
    },
    {
        title: "Step two",
        ongoing: true
    },
    {
        title: "Step three",
        selected: true
    },
    {
        title: "Step four"
    }
];

const StepperContent = ({ completed, subtitle, title, index }) => (
    <>
        {completed ? <><div className="material-icons stepper-icon" >check</div>{ "\n" }</> : <div className="stepper-number">{index + 1}</div>}
        {title}
        {subtitle ? <>{ "\n" }<div className="stepper-sub-title">{subtitle}</div></> : null}
    </>
);

const RenderSteps = ({ steps }) => (
    <>
        {steps.map(({ title, subtitle, completed, ongoing, selected, clickable }, i) => (
            <li key={i} className={classnames(completed ? "stepper-completed" : null,
                ongoing ? "stepper-ongoing" : null,
                selected ? "stepper-selected" : null)}>
                { clickable ?
                    <>{ "\n" } <a><StepperContent completed={completed} subtitle={subtitle} title={title} index={i}/></a> { "\n" } </>
                    :
                    <StepperContent completed={completed} subtitle={subtitle} title={title} index={i}/>
                }
            </li>
        ))}
    </>
);

const RenderStepperNav = ({ steps }) => {
    if (!steps) {
        return null;
    }

    const clickableItem = steps.some(item => item.clickable);

    const arrowLeft = clickableItem ? <div className="material-icons stepper-nav-left">keyboard_arrow_left</div> : null;
    const arrowRight = clickableItem ? <div className="material-icons stepper-nav-right">keyboard_arrow_right</div> : null;

    return <>
        <div className="stepper-responsive">
            { arrowLeft }
            <div className="stepper-responsive-text">Step 3</div>
            { arrowRight }
        </div>
    </>;
};

const Stepper = ({ steps, vertical }) => {
    const stepsClasses = classnames(
        "stepper",
        vertical ? "stepper-vertical" : null,
    );

    return (<div className={stepsClasses}>
        <ol>
            {steps ? <RenderSteps steps={steps} /> : <RenderSteps steps={DefaultStepper} />}
        </ol>
        <RenderStepperNav steps={steps} />
    </div>);
};

Stepper.propTypes = {
    steps: PropTypes.array,
    vertical: PropTypes.bool,
    clickable: PropTypes.bool
};

export default Stepper;

// exporting for testing purposes
export { DefaultStepper };
