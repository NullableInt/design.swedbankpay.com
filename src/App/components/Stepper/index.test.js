import React from "react";
import { shallow, mount } from "enzyme";

import Stepper from "./index";

describe("Component: Stepper", () => {
    const steps = [
        {
            title: "Step one",
            completed: true,
            clickable: true
        },
        {
            title: "Step two",
            ongoing: true,
            clickable: true
        },
        {
            title: "Step three",
            clickable: true
        },
        {
            title: "Step four",
            selected: true
        },
        {
            title: "Step Five"
        },
        {
            title: "Step Six"
        }
    ];

    const simpleSteps = [
        {
            title: "one step"
        },
        {
            title: "two steps"
        },
        {
            title: "three steps for testing"
        }
    ];

    it("is defined", () => {
        expect(Stepper).toBeDefined();
    });

    it("renders with default items if no props are provided", () => {
        const wrapper = shallow(<Stepper />);

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.html()).toContain("stepper");
    });

    it("renders with given items if prop Stepper is given", () => {
        const wrapper = mount(<Stepper steps={simpleSteps} />);

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.html()).toContain("three steps for testing");
    });

    it("renders in vertical mode if prop vertical is provided", () => {
        const wrapper = shallow(<Stepper vertical />);

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.html()).toContain("stepper-vertical");
    });

    it("renders clickable Stepper", () => {
        const wrapper = mount(<Stepper steps={steps} />);

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find(".stepper a").length).toBeGreaterThan(0);
    });
});
