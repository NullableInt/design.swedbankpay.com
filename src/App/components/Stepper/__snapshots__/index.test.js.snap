// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Stepper renders clickable Stepper 1`] = `
<Stepper
  steps={
    Array [
      Object {
        "clickable": true,
        "completed": true,
        "title": "Step one",
      },
      Object {
        "clickable": true,
        "ongoing": true,
        "title": "Step two",
      },
      Object {
        "clickable": true,
        "title": "Step three",
      },
      Object {
        "selected": true,
        "title": "Step four",
      },
      Object {
        "title": "Step Five",
      },
      Object {
        "title": "Step Six",
      },
    ]
  }
>
  <div
    className="stepper"
  >
    <ol>
      <RenderSteps
        steps={
          Array [
            Object {
              "clickable": true,
              "completed": true,
              "title": "Step one",
            },
            Object {
              "clickable": true,
              "ongoing": true,
              "title": "Step two",
            },
            Object {
              "clickable": true,
              "title": "Step three",
            },
            Object {
              "selected": true,
              "title": "Step four",
            },
            Object {
              "title": "Step Five",
            },
            Object {
              "title": "Step Six",
            },
          ]
        }
      >
        <li
          className="stepper-completed"
          key="0"
        >
          

           
          <a>
            <StepperContent
              completed={true}
              index={0}
              title="Step one"
            >
              <div
                className="material-icons stepper-icon"
              >
                check
              </div>
              

              Step one
            </StepperContent>
          </a>
           
          

           
        </li>
        <li
          className="stepper-ongoing"
          key="1"
        >
          

           
          <a>
            <StepperContent
              index={1}
              title="Step two"
            >
              <div
                className="stepper-number"
              >
                2
              </div>
              Step two
            </StepperContent>
          </a>
           
          

           
        </li>
        <li
          className=""
          key="2"
        >
          

           
          <a>
            <StepperContent
              index={2}
              title="Step three"
            >
              <div
                className="stepper-number"
              >
                3
              </div>
              Step three
            </StepperContent>
          </a>
           
          

           
        </li>
        <li
          className="stepper-selected"
          key="3"
        >
          <StepperContent
            index={3}
            title="Step four"
          >
            <div
              className="stepper-number"
            >
              4
            </div>
            Step four
          </StepperContent>
        </li>
        <li
          className=""
          key="4"
        >
          <StepperContent
            index={4}
            title="Step Five"
          >
            <div
              className="stepper-number"
            >
              5
            </div>
            Step Five
          </StepperContent>
        </li>
        <li
          className=""
          key="5"
        >
          <StepperContent
            index={5}
            title="Step Six"
          >
            <div
              className="stepper-number"
            >
              6
            </div>
            Step Six
          </StepperContent>
        </li>
      </RenderSteps>
    </ol>
    <RenderStepperNav
      steps={
        Array [
          Object {
            "clickable": true,
            "completed": true,
            "title": "Step one",
          },
          Object {
            "clickable": true,
            "ongoing": true,
            "title": "Step two",
          },
          Object {
            "clickable": true,
            "title": "Step three",
          },
          Object {
            "selected": true,
            "title": "Step four",
          },
          Object {
            "title": "Step Five",
          },
          Object {
            "title": "Step Six",
          },
        ]
      }
    >
      <div
        className="stepper-responsive"
      >
        <div
          className="material-icons stepper-nav-left"
        >
          keyboard_arrow_left
        </div>
        <div
          className="stepper-responsive-text"
        >
          Step 3
        </div>
        <div
          className="material-icons stepper-nav-right"
        >
          keyboard_arrow_right
        </div>
      </div>
    </RenderStepperNav>
  </div>
</Stepper>
`;

exports[`Component: Stepper renders in vertical mode if prop vertical is provided 1`] = `
<div
  className="stepper stepper-vertical"
>
  <ol>
    <RenderSteps
      steps={
        Array [
          Object {
            "completed": true,
            "subtitle": "24.12.17 12:10",
            "title": "Step one",
          },
          Object {
            "ongoing": true,
            "title": "Step two",
          },
          Object {
            "selected": true,
            "title": "Step three",
          },
          Object {
            "title": "Step four",
          },
        ]
      }
    />
  </ol>
  <RenderStepperNav />
</div>
`;

exports[`Component: Stepper renders with default items if no props are provided 1`] = `
<div
  className="stepper"
>
  <ol>
    <RenderSteps
      steps={
        Array [
          Object {
            "completed": true,
            "subtitle": "24.12.17 12:10",
            "title": "Step one",
          },
          Object {
            "ongoing": true,
            "title": "Step two",
          },
          Object {
            "selected": true,
            "title": "Step three",
          },
          Object {
            "title": "Step four",
          },
        ]
      }
    />
  </ol>
  <RenderStepperNav />
</div>
`;

exports[`Component: Stepper renders with given items if prop Stepper is given 1`] = `
<Stepper
  steps={
    Array [
      Object {
        "title": "one step",
      },
      Object {
        "title": "two steps",
      },
      Object {
        "title": "three steps for testing",
      },
    ]
  }
>
  <div
    className="stepper"
  >
    <ol>
      <RenderSteps
        steps={
          Array [
            Object {
              "title": "one step",
            },
            Object {
              "title": "two steps",
            },
            Object {
              "title": "three steps for testing",
            },
          ]
        }
      >
        <li
          className=""
          key="0"
        >
          <StepperContent
            index={0}
            title="one step"
          >
            <div
              className="stepper-number"
            >
              1
            </div>
            one step
          </StepperContent>
        </li>
        <li
          className=""
          key="1"
        >
          <StepperContent
            index={1}
            title="two steps"
          >
            <div
              className="stepper-number"
            >
              2
            </div>
            two steps
          </StepperContent>
        </li>
        <li
          className=""
          key="2"
        >
          <StepperContent
            index={2}
            title="three steps for testing"
          >
            <div
              className="stepper-number"
            >
              3
            </div>
            three steps for testing
          </StepperContent>
        </li>
      </RenderSteps>
    </ol>
    <RenderStepperNav
      steps={
        Array [
          Object {
            "title": "one step",
          },
          Object {
            "title": "two steps",
          },
          Object {
            "title": "three steps for testing",
          },
        ]
      }
    >
      <div
        className="stepper-responsive"
      >
        <div
          className="stepper-responsive-text"
        >
          Step 3
        </div>
      </div>
    </RenderStepperNav>
  </div>
</Stepper>
`;
